builtin/credential/approle/path_login.go

```diff
@@ -51,14 +53,186 @@ func (b *backend) pathLoginUpdateAliasLookahead(ctx context.Context, req *logica
 // Returns the Auth object indicating the authentication and authorization information
 // if the credentials provided are validated by the backend.
 func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-	role, roleName, metadata, _, userErr, intErr := b.validateCredentials(ctx, req, data)
-	switch {
-	case intErr != nil:
-		return nil, errwrap.Wrapf("failed to validate credentials: {{err}}", intErr)
-	case userErr != nil:
-		return logical.ErrorResponse(fmt.Sprintf("failed to validate credentials: %v", userErr)), nil
-	case role == nil:
-		return logical.ErrorResponse("failed to validate credentials; could not find role"), nil
+
+	// RoleID must be supplied during every login
+	roleID := strings.TrimSpace(data.Get("role_id").(string))
+	if roleID == "" {
+		return logical.ErrorResponse("missing role_id"), nil
+	}
+
+	// Look for the storage entry that maps the roleID to role
+	roleIDIndex, err := b.roleIDEntry(ctx, req.Storage, roleID)
+	if err != nil {
+		return nil, err
+	}
+	if roleIDIndex == nil {
+		return logical.ErrorResponse(fmt.Sprintf("invalid role_id %q", roleID)), nil
+	}
+
+	roleName := roleIDIndex.Name
+
+	roleLock := b.roleLock(roleName)
+	roleLock.RLock()
+	defer roleLock.RUnlock()
```

Comment:
Do you actually need the lock after the roleEntry call? If not, can you just manually unlock instead of deferring until the end? That way you're not holding on to the role lock while also contending for the read/write locks for the secret ID.